### PR TITLE
[wrong branch] Add bucket notion

### PR DESCRIPTION
**Pull request by leplatrem**
(Based on #26)

**Do not merge before** #26 and #22 

r? = @n1k0, especially options handling.
(review by commit will be easier)

### CODE REVIEW

File: src/api.js [hunk 1]
```diff
@@ -40,47 +42,58 @@ export default class Api {
...
-    return {
+    var endpoints = {
       root:           () => root,
       batch:          () => `${root}/batch`,
-      collection: (coll) => `${root}/collections/${coll}/records`,
-      record: (coll, id) => `${this.endpoints(options).collection(coll)}/${id}`,
+      bucket: (bucket) => `${root}/buckets/${bucket}`,
+      collection: (bucket, coll) => `${endpoints.bucket(bucket)}/collections/${coll}`,
+      records: (bucket, coll) => `${endpoints.collection(bucket, coll)}/records`,
+      record: (bucket, coll, id) => `${endpoints.records(bucket, coll)}/${id}`,
```
Comment (n1k0) [inline src/api.js]: Nit: indentation/alignment, but you're probably already aware of that :)

File: src/api.js [hunk 2]
```diff
@@ -40,47 +42,58 @@ export default class Api {
...
     };
+    return endpoints;
```
Comment (n1k0) [inline src/api.js]: Nit: here I can't really see the benefit of not returning the object directly. Hint?

Comment (n1k0) [inline src/api.js]: Hmm, ok reference to the object used ithin template strings; scratch previous comment.

File: src/api.js (hunk 1 above)
Comment (leplatrem) [inline src/api.js]: Yes indeed you're right :) also I didn't respect max length of 80 chars :(

File: src/api.js [hunk 3]
```diff
@@ -40,47 +42,58 @@ export default class Api {
...
    */
-  fetchChangesSince(collName, lastModified=null, options={headers: {}}) {
+  fetchChangesSince(bucketName, collName, options={lastModified: null, headers: {}}) {
     var newLastModified;
-    var queryString = "?" + (lastModified ? "_since=" + lastModified : "");
-    return fetch(this.endpoints().collection(collName) + queryString, {
-      headers: Object.assign({}, DEFAULT_REQUEST_HEADERS, {
-        "If-Modified-Since": lastModified ? String(lastModified) : "0"
-      }, options.headers)
+    var recordsUrl = this.endpoints().records(bucketName, collName);
```
Comment (n1k0) [inline src/api.js]: Nit: this sould be a `const`

File: src/api.js [hunk 4]
```diff
@@ -40,47 +42,58 @@ export default class Api {
...
+    var headers = Object.assign({}, DEFAULT_REQUEST_HEADERS, options.headers);
+
+    if (options.lastModified) {
+      queryString = "?_since=" + options.lastModified;
+      headers["If-None-Match"] = quote(options.lastModified);
+    }
+
+    return fetch(recordsUrl + queryString, {
+      headers: headers
     })
```
Comment (n1k0) [inline src/api.js]: Nit: this is probably as legible on a single line

File: src/api.js [hunk 5]
```diff
@@ -110,15 +125,23 @@ export default class Api {
       method: "POST",
       headers: DEFAULT_REQUEST_HEADERS,
       body: JSON.stringify({
-        defaults: { headers },
+        defaults: {headers: headers},
```
Comment (n1k0) [inline src/api.js]: Why this change?

File: src/api.js [hunk 6]
```diff
@@ -110,15 +125,23 @@ export default class Api {
...
-            "If-Unmodified-Since": String(record.last_modified || "0")
-          } : {};
+          const body = isDeletion ? undefined : { data: cleanRecord(record) };
+          const headers = {};
+          if (safe) {
+            if (record.last_modified) {
+              // Safe replace.
+              headers["If-Match"] = quote(record.last_modified);
+            }
+            else if (!isDeletion) {
```
Comment (n1k0) [inline src/api.js]: Nit: I think we usually prefered having the whole `else (if)` statement + brackets on a single line

File: src/api.js [hunk 7]
```diff
@@ -135,7 +158,7 @@ export default class Api {
           // TODO: handle 409 when unicity rule is violated (ex. POST with
           // existing id, unique field, etc.)
           if (response.status && response.status >= 200 && response.status < 400) {
-            results.published.push(response.body);
+            results.published.push(response.body.data);
```
Comment (n1k0) [inline src/api.js]: Yes!

File: src/collection.js [hunk 8]
```diff
@@ -36,7 +36,8 @@ export class SyncResultObject {
 }
 
 export default class Collection {
-  constructor(name, api) {
+  constructor(bucket, name, api) {
```
Comment (n1k0) [inline src/collection.js]: Maybe take this opportunity to add some jsdoc here

File: src/collection.js [hunk 9]
```diff
@@ -463,8 +472,9 @@ export default class Collection {
    * @return {Promise}
    */
   pullChanges(syncResultObject, options) {
```
Comment (n1k0) [inline src/collection.js]: You should define a default value for the `options` arg (`{}`)

File: src/collection.js [hunk 10]
```diff
@@ -486,9 +499,7 @@ export default class Collection {
             return this.delete(record.id, {virtual: false});
           })),
           // Send batch update requests
-          this.api.batch(this.name, toSync, options.headers, {
-            safe: options.strategy === Collection.SERVER_WINS
-          })
+          this.api.batch(this.bucket, this.name, toSync, options)
```
Comment (n1k0) [inline src/collection.js]: Much cleaner.

File: src/index.js [hunk 11]
```diff
@@ -17,9 +17,10 @@ export default class Cliquetis {
       throw new Error("missing collection name");
 
     const api = new Api(this._options.remote || "http://0.0.0.0:8888/v0");
+    const bucket = this._options.bucket || "default";
```
Comment (n1k0) [inline src/index.js]: Maybe we could define and use a `DEFAULT_BUCKET` const here?

File: src/utils.js [hunk 12]
```diff
@@ -23,3 +23,17 @@ export function attachFakeIDBSymbolsTo(obj) {
   })
   obj.indexedDB = require("fake-indexeddb");
 }
+
+/**
+ * Returns the specified string with double quotes.
+ */
```
Comment (n1k0) [inline src/utils.js]: Nit: @param & @return missing

File: src/utils.js [hunk 13]
```diff
@@ -23,3 +23,17 @@ export function attachFakeIDBSymbolsTo(obj) {
...
+export function quote(str) {
+  return `"${str}"`;
+}
+
+/**
+ * Trim double quotes from specified string.
+ */
```
Comment (n1k0) [inline src/utils.js]: Nitto

File: test/api_test.js [hunk 14]
```diff
@@ -58,7 +58,7 @@ describe("Api", () => {
     describe("full URL", () => {
       var endpoints;
 
-      beforeEach(() => endpoints = api.endpoints({fullUrl: true}))
+      beforeEach(() => endpoints = api.endpoints({ fullUrl: true }))
```
Comment (n1k0) [inline test/api_test.js]: We should make a decision and stick to it =)

File: test/api_test.js [hunk 15]
```diff
@@ -118,28 +138,27 @@ describe("Api", () => {
...
       sinon.assert.calledWithMatch(fetch, /\?_since=42/, {
-        headers: {"If-Modified-Since": "42"}
+        headers: { "If-None-Match": '"42"' }
       });
     });
 
     it("should resolve with a result object", () => {
       sandbox.stub(root, "fetch").returns(
-        fakeServerResponse(200, {items: []}, {"Last-Modified": 41}));
+        fakeServerResponse(200, { data: [] }, { "ETag": '"41"' }));
```
Comment (n1k0) [inline test/api_test.js]: How about reusing `quote()` here?

File: test/api_test.js [hunk 16]
```diff
@@ -217,36 +237,43 @@ describe("Api", () => {
...
         it("should map create & update requests", () => {
           expect(requestBody.requests[0]).eql({
-            "body": {
-              "id": 1,
-              "title": "foo",
+            body: {
+              data: { id: 1, title: "foo" },
             },
-            headers: {"If-Unmodified-Since": "42"},
+            headers: { "If-Match": '"42"' },
```
Comment (n1k0) [inline test/api_test.js]: `quote(42)` should work

File: test/api_test.js [hunk 17]
```diff
@@ -217,36 +237,43 @@ describe("Api", () => {
...
           });
         });
 
-        it("should map batch delete requests for synced records", () => {
+        it("should map batch update requests for synced records", () => {
           expect(requestBody.requests[0]).eql({
-            path: "/v0/collections/articles/records/1",
+            path: "/v0/buckets/blog/collections/articles/records/1",
+            method: "PUT",
+            headers: { "If-Match": '"42"' },
```
Comment (n1k0) [inline test/api_test.js]: `quote(42)` should work here too

File: test/api_test.js [hunk 18]
```diff
@@ -257,12 +284,12 @@ describe("Api", () => {
...
           sandbox.stub(root, "fetch").returns(Promise.resolve({status: 200}));
-          api.batch("articles", operations);
+          api.batch("blog", "articles", operations);
           requests = JSON.parse(fetch.getCall(0).args[1].body).requests;
         });
 
-        it("should send If-Unmodified-Since headers", () => {
-          expect(requests[0].headers).eql({"If-Unmodified-Since": "42"});
+        it("should send If-Match headers", () => {
+          expect(requests[0].headers).eql({"If-Match": '"42"'});
```
Comment (n1k0) [inline test/api_test.js]: Ditto

File: src/api.js [hunk 19]
```diff
@@ -91,13 +104,15 @@ export default class Api {
...
    *
+   * @param  {String} bucketName   The bucket name.
    * @param  {String} collName The collection name.
    * @param  {Array}  records  The list of record updates to send.
-   * @param  {Object} headers  Headers to attach to each update request.
    * @param  {Object} options  Options.
    * @return {Promise}
    */
-  batch(collName, records, headers={}, options={safe: true}) {
+  batch(bucketName, collName, records, options={headers: {}, safe: true}) {
```
Comment (n1k0) [inline src/api.js]: I'm now wondering if `bucketName` shouldn't be an option; users may be most likely wanting the default bucket most of the time… What do you think?

**Comment by n1k0** [comment]
Minor nits here and there, this looks super good overall. My only open question is should we actually the bucket name as a mandatory arg in Api methods?

**Comment by leplatrem** [comment]
Closing in favor of #19
